template.tpl

```diff
@@ -0,0 +1,1704 @@
+  const returnObj = payload;
+  if (tagConfig.includeCommonEventProperties) {
+    COMMON_EVENT_PROPS.forEach((prop) => {
+      addProperty(prop, evData[prop], tagConfig, NEST.commonEvent, returnObj);
+    });
+  }
+  if (tagConfig.includeCommonUserProperties) {
+    const uKey = 'user_data';
+    if (evData[uKey]) {
+      addProperty(uKey, evData[uKey], tagConfig, NEST.userData, returnObj);
+    }
+  }
+  if (tagConfig.eventMappingRules) {
+    tagConfig.eventMappingRules.forEach((row) => {
+      const setVal = getEventData(row.key);
+      const mappedKey = row.mappedKey || row.key;
+      setFromPath(mappedKey, setVal, tagConfig, returnObj);
+    });
+  }
+  return returnObj;
+};
+
+/*
+ * Adds key-value pairs to the payload according to tag configuration.
+ *
+ * @param evData {Object} - the client event data
+ * @param tagConfig {Object} - the tag configuration
+ * @param payload {Object} - the payload object
+ * @returns - the mutated payload with the selected key-value pairs added
+ */
+const addRequestData = (tagConfig, payload) => {
+  const returnObj = payload;
+  if (tagConfig.additionalRequestData) {
+    tagConfig.additionalRequestData.forEach((row) => {
+      setFromPath(row.key, row.value, tagConfig, returnObj);
+    });
+  }
+  return returnObj;
+};
+
+/*
+ * Given the event data and the tag configuration, constructs the request body.
+ */
+const mkRequestPayload = (evData, tagConfig) => {
+  if (tagConfig.includeAll) {
+    return addRequestData(tagConfig, evData);
+  }
+
+  const target = {};
+  const requestPayload = cleanObject(
+    addRequestData(
+      tagConfig,
+      addEventMappings(
+        evData,
+        tagConfig,
+        parseSnowplowEvent(evData, tagConfig, target)
+      )
+    )
+  );
+
+  return tagConfig.inArray ? [requestPayload] : requestPayload;
+};
+
+/*
+ * Creates the HTTP request options according to the tag configuration.
+ */
+const mkRequestOptions = (tagConfig) => {
+  // requestTimeout input ensured by its validation rules
+  const timeout = makeInteger(tagConfig.requestTimeout);
+  const options = {
+    method: tagConfig.requestMethod.toUpperCase(),
+    timeout: timeout,
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  };
+
+  if (tagConfig.headers) {
+    tagConfig.headers.forEach((h) => {
+      options.headers[h.key] = h.value;
+    });
+  }
+
+  return options;
+};
+
+// Main
+const url = data.url;
+const eventData = getAllEventData();
+
+const requestOptions = mkRequestOptions(data);
+const requestPayload = mkRequestPayload(eventData, data);
+
+if (requestPayload) {
+  sendHttpRequest(
+    url,
+    (statusCode, headers, body) => {
+      if (statusCode >= 200 && statusCode < 300) {
+        data.gtmOnSuccess();
+        return;
+      }
+      data.gtmOnFailure();
+    },
+    requestOptions,
+    JSON.stringify(requestPayload)
+  );
+} else {
+  fail('Unable to build request from supplied Client event.');
```

Comment:
I _think_ you need to `return fail(...)` here, as `fail()` returns the gtm failure functions return value? Not entirely sure but seems safer.